tests/posix/pthread_pressure/src/main.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2023, Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <pthread.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/ztest.h>
+
+#define STACK_SIZE K_THREAD_STACK_LEN(CONFIG_TEST_STACK_SIZE)
+
+/* 32 threads is mainly a limitation of find_lsb_set() */
+#define NUM_THREADS MIN(32, MIN(CONFIG_TEST_NUM_CPUS, CONFIG_MAX_PTHREAD_COUNT))
+
+LOG_MODULE_REGISTER(pthread_pressure, LOG_LEVEL_INF);
+
+/* bitmask of available threads */
+static atomic_t free_threads = ATOMIC_INIT(BIT_MASK(NUM_THREADS));
+
+/* array of thread stacks */
+static K_THREAD_STACK_ARRAY_DEFINE(thread_stacks, NUM_THREADS, STACK_SIZE);
+
+/* array of pthread_t */
+static pthread_t pthreads[NUM_THREADS];
+static pthread_attr_t pthread_attrs[NUM_THREADS];
+static size_t counters[NUM_THREADS];
+
+/* no-op thread */
+static void *pthread_fun(void *arg)
+{
+	const size_t i = (size_t)arg;
+
+	ARG_UNUSED(arg);
+	atomic_set_bit(&free_threads, i);
+	LOG_DBG("thread %zu exiting..", i);
+
+	return NULL;
+}
+
+ZTEST(pthread_pressure, test_pthread_create_join)
+{
+	int ret;
+	size_t i;
+	uint64_t now;
+	uint32_t maybe_free_threads;
+	const uint64_t end =
+		sys_clock_timeout_end_calc(K_MSEC(MSEC_PER_SEC * CONFIG_TEST_DURATION_S));
+
+	LOG_INF("NUM_THREADS: %u", NUM_THREADS);
+	LOG_INF("TEST_NUM_CPUS: %u", CONFIG_TEST_NUM_CPUS);
+	LOG_INF("TEST_DURATION_S: %u", CONFIG_TEST_DURATION_S);
+	LOG_INF("TEST_DELAY_US: %u", CONFIG_TEST_DELAY_US);
+
+	for (i = 0; i < NUM_THREADS; ++i) {
+		zassert_ok(pthread_attr_init(&pthread_attrs[i]));
+		zassert_ok(
+			pthread_attr_setstack(&pthread_attrs[i], &thread_stacks[i][0], STACK_SIZE));
+		atomic_clear_bit(&free_threads, i);
+		ret = pthread_create(&pthreads[i], &pthread_attrs[i], pthread_fun, (void *)i);
+		zassert_ok(ret, "unable to spawn thread %zu: %d", i, ret);
+		LOG_DBG("spawned thread %zu", i);
+	}
+
+	do {
+		maybe_free_threads = (uint32_t)atomic_get(&free_threads);
+		LOG_DBG("free_threads: %x", maybe_free_threads);
+		while (maybe_free_threads != 0) {
+			i = find_lsb_set(maybe_free_threads) - 1;
```

Comment:
This looks like it will re-use the LSB value and possibly never hit other threads? Maybe `pthread_fun` should use `k_busy_wait` to spin the CPU for a while (random?) so that threads are actually active for some time?

Reply:
> This looks like it will re-use the LSB value and possibly never hit other threads? Maybe `pthread_fun` should use `k_busy_wait` to spin the CPU for a while (random?) so that threads are actually active for some time?

The point of the test was to not to introduce any unnecessary delays between creation and join.

The only necessary delay was to ensure that the thread is spawned before joining as soon as possible.

Technically speaking, this could also be solved by introducing another state.

Reply:
I don't think we need any more state in the pthread implementation; we just need to ensure that serialized calls between `k_thread_create` and `k_thread_join` are always valid, no matter what the created thread is doing.

Reply:
As it turns out, even those calls outside of pthreads are racey 😞